File: python/src/tdcv2/output/parquet/thrift.py
"""Thrift's compact protocol, write side only.

Parquet keeps its page headers and its entire footer in this encoding, so a file cannot be produced
without it. Small and completely specified — and unforgiving: one wrong byte and no reader will
open the file, with nothing to say about which byte. That is why it lives on its own and is checked
against known bytes.
"""

from __future__ import annotations

# Compact type ids. A boolean carries its value in the field header rather than after it.
BOOLEAN_TRUE = 1
BOOLEAN_FALSE = 2
BYTE = 3
I16 = 4
I32 = 5
I64 = 6
DOUBLE = 7
BINARY = 8
LIST = 9
SET = 10
MAP = 11
STRUCT = 12


def varint(value: int) -> bytes:
    """Unsigned LEB128: seven bits per byte, the top bit meaning "more follows"."""
    if value < 0:
        raise ValueError("varint must be non-negative")
    out = bytearray()
    v = value
    while True:
        b = v & 0x7F
        v >>= 7
        if v > 0:
            b |= 0x80
        out.append(b)
        if v == 0:
            return bytes(out)


def zigzag32(value: int) -> int:
    """A signed 32-bit value folded onto an unsigned one, so small magnitudes stay short."""
    return ((value << 1) ^ (value >> 31)) & 0xFFFFFFFF


def zigzag64(value: int) -> int:
    return ((value << 1) ^ (value >> 63)) & 0xFFFFFFFFFFFFFFFF


class Writer:
    __slots__ = ("_last_field_id", "_out", "_stack")

    def __init__(self) -> None:
        self._out = bytearray()
        # Field ids are written as a delta from the previous field of the same struct.
        self._last_field_id = 0
        self._stack: list[int] = []

    def bytes(self) -> bytes:
        return bytes(self._out)

    def __len__(self) -> int:
        """How many bytes so far — what page and footer offsets are filled in from."""
        return len(self._out)

    def struct_begin(self) -> None:
        self._stack.append(self._last_field_id)
        self._last_field_id = 0

    def struct_end(self) -> None:
        self._out.append(0x00)  # struct stop
        self._last_field_id = self._stack.pop() if self._stack else 0

    def field_begin(self, field_id: int, type_: int) -> None:
        """The short form when the id delta fits in four bits, the long form otherwise."""
        delta = field_id - self._last_field_id
        if 0 < delta <= 15:
            self._out.append((delta << 4) | type_)
        else:
            self._out.append(type_)
            self._out += varint(zigzag32(field_id))
        self._last_field_id = field_id

    def bool(self, field_id: int, value: bool) -> None:
        """A boolean has no value bytes: true and false are two different field types."""
        self.field_begin(field_id, BOOLEAN_TRUE if value else BOOLEAN_FALSE)

    def i8(self, field_id: int, value: int) -> None:
        """Thrift's ``i8`` — one raw byte, NOT zigzagged the way i16/i32/i64 are.

        ``LogicalType.IntType.bitWidth`` is declared i8, and writing it as an i32 would shift every
        field after it by a byte.
        """
        self.field_begin(field_id, BYTE)
        self._out.append(value & 0xFF)

    def i32(self, field_id: int, value: int) -> None:
        self.field_begin(field_id, I32)
        self._out += varint(zigzag32(value))

    def i64(self, field_id: int, value: int) -> None:
        self.field_begin(field_id, I64)
        self._out += varint(zigzag64(value))

    def binary(self, field_id: int, value: bytes) -> None:
        self.field_begin(field_id, BINARY)
        self._out += varint(len(value))
        self._out += value

    def string(self, field_id: int, value: str) -> None:
        self.binary(field_id, value.encode("utf-8"))

    def list_begin(self, field_id: int, element_type: int, size: int) -> None:
        """Open a list field.

        Its elements follow with the ``list_*`` writers and carry no field headers of their own; a
        list of structs uses :meth:`struct_begin` and :meth:`struct_end`.
        """
        self.field_begin(field_id, LIST)
        if size < 15:
            self._out.append((size << 4) | element_type)
        else:
            self._out.append((0x0F << 4) | element_type)
            self._out += varint(size)

    def list_i32(self, value: int) -> None:
        self._out += varint(zigzag32(value))

    def list_i64(self, value: int) -> None:
        self._out += varint(zigzag64(value))

    def list_string(self, value: str) -> None:
        encoded = value.encode("utf-8")
        self._out += varint(len(encoded))
        self._out += encoded

File: csharp/Tdcv2/Output/Parquet/Thrift.cs
using System.Text;

namespace Tdcv2.Output.Parquet;

/// <summary>
/// Thrift's compact protocol, write side only.
/// </summary>
/// <remarks>
/// Parquet keeps its page headers and its entire footer in this encoding, so a file cannot be
/// produced without it. Small and completely specified — and unforgiving: one wrong byte and no
/// reader will open the file, with nothing to say about which byte. That is why it lives on its own
/// and is checked against known bytes.
/// </remarks>
public sealed class Thrift
{
    /// <summary>Compact type ids. A boolean carries its value in the field header rather than after it.</summary>
    public const int BooleanTrue = 1;

    public const int BooleanFalse = 2;
    public const int Byte = 3;
    public const int I16Type = 4;
    public const int I32Type = 5;
    public const int I64Type = 6;
    public const int DoubleType = 7;
    public const int Binary = 8;
    public const int ListType = 9;
    public const int SetType = 10;
    public const int MapType = 11;
    public const int StructType = 12;

    private readonly MemoryStream _out = new();

    /// <summary>Field ids are written as a delta from the previous field of the same struct.</summary>
    private int _lastFieldId;

    private readonly Stack<int> _stack = new();

    public byte[] Bytes() => _out.ToArray();

    /// <summary>How many bytes so far — what page and footer offsets are filled in from.</summary>
    public int Length => (int)_out.Length;

    /// <summary>Unsigned LEB128: seven bits per byte, the top bit meaning "more follows".</summary>
    public static byte[] Varint(long value)
    {
        if (value < 0)
        {
            throw new ArgumentException("varint must be non-negative");
        }

        var buffer = new List<byte>();
        ulong v = (ulong)value;
        do
        {
            int b = (int)(v & 0x7f);
            v >>= 7;
            if (v > 0)
            {
                b |= 0x80;
            }

            buffer.Add((byte)b);
        }
        while (v > 0);

        return buffer.ToArray();
    }

    /// <summary>Fold a signed 32-bit value onto an unsigned one so small magnitudes stay short.</summary>
    public static long Zigzag32(int value) => (long)(uint)((value << 1) ^ (value >> 31));

    public static long Zigzag64(long value) => (value << 1) ^ (value >> 63);

    private void Raw(int b) => _out.WriteByte((byte)(b & 0xff));

    private void RawBytes(byte[] bytes) => _out.Write(bytes, 0, bytes.Length);

    public void StructBegin()
    {
        _stack.Push(_lastFieldId);
        _lastFieldId = 0;
    }

    public void StructEnd()
    {
        Raw(0x00); // struct stop
        _lastFieldId = _stack.Count == 0 ? 0 : _stack.Pop();
    }

    /// <summary>A field header: the short form when the id delta fits in four bits, the long form otherwise.</summary>
    public void FieldBegin(int id, int type)
    {
        int delta = id - _lastFieldId;
        if (delta > 0 && delta <= 15)
        {
            Raw((delta << 4) | type);
        }
        else
        {
            Raw(type);
            RawBytes(Varint(Zigzag32(id)));
        }

        _lastFieldId = id;
    }

    /// <summary>A boolean has no value bytes: true and false are two different field types.</summary>
    public void Bool(int id, bool value) =>
        FieldBegin(id, value ? BooleanTrue : BooleanFalse);

    /// <summary>
    /// Thrift's <c>i8</c> — one raw byte, NOT zigzagged the way i16/i32/i64 are.
    /// </summary>
    /// <remarks>
    /// <c>LogicalType.IntType.bitWidth</c> is declared i8, and writing it as an i32 would shift every
    /// field after it by a byte.
    /// </remarks>
    public void I8(int id, int value)
    {
        FieldBegin(id, Byte);
        Raw(value);
    }

    public void I32(int id, int value)
    {
        FieldBegin(id, I32Type);
        RawBytes(Varint(Zigzag32(value)));
    }

    public void I64(int id, long value)
    {
        FieldBegin(id, I64Type);
        RawBytes(Varint(Zigzag64(value)));
    }

    public void BinaryField(int id, byte[] value)
    {
        FieldBegin(id, Binary);
        RawBytes(Varint(value.Length));
        RawBytes(value);
    }

    public void String(int id, string value) =>
        BinaryField(id, Encoding.UTF8.GetBytes(value));

    /// <summary>
    /// Open a list field. Its elements follow with the <c>List*</c> writers and carry no field
    /// headers of their own; a list of structs uses <see cref="StructBegin"/> and
    /// <see cref="StructEnd"/>.
    /// </summary>
    public void ListBegin(int id, int elementType, int size)
    {
        FieldBegin(id, ListType);
        if (size < 15)
        {
            Raw((size << 4) | elementType);
        }
        else
        {
            Raw((0x0f << 4) | elementType);
            RawBytes(Varint(size));
        }
    }

    public void ListI32(int value) => RawBytes(Varint(Zigzag32(value)));

    public void ListI64(long value) => RawBytes(Varint(Zigzag64(value)));

    public void ListString(string value)
    {
        byte[] bytes = Encoding.UTF8.GetBytes(value);
        RawBytes(Varint(bytes.Length));
        RawBytes(bytes);
    }
}
